src/mem/largealloc.h

```diff
@@ -207,15 +209,26 @@ namespace snmalloc
       lazy_decommit_guard.clear();
     }
 
-    /***
-     * Method for callback object to perform lazy decommit.
-     */
-    static void process(PalNotificationObject* p)
+    class LowMemoryNotificationObject : public PalNotificationObject
     {
-      // Unsafe downcast here. Don't want vtable and RTTI.
-      auto self = reinterpret_cast<MemoryProviderStateMixin<PAL>*>(p);
-      self->lazy_decommit();
-    }
+      MemoryProviderStateMixin<PAL>* memory_provider;
+
+      /***
+       * Method for callback object to perform lazy decommit.
+       */
+      static void process(PalNotificationObject* p)
+      {
+        // Unsafe downcast here. Don't want vtable and RTTI.
+        auto self = reinterpret_cast<LowMemoryNotificationObject*>(p);

```

Comment:
Do you anticipate `PalNotificationObject`s that don't need to refer to the `memory_provider`?  If not, maybe just move the field up into the parent type?  Or maybe templatize `PalNotificationObject` on the type of state it's carrying?

Reply:
Not easily.  As the Pal doesn't know about memory providers.  We need the list to carry types of this, so really we need existential types.  In C++ that really means virtual, but generally we have tried to stay away from C++ vtables and RTTI in the allocator.